### ISSUE TITLE
Dollar sign before config path or custom variables in cms page content makes listing crash

### ISSUE THREAD

**Issue by fxdegroot** [magento/magento2#10501]
### Preconditions

1. Magento 2 CE 2.1.7 & Magento 2 CE 2.1.8
2. PHP 7.0.19

### Steps to reproduce

1. In the admin, edit any CMS page
2. From the content section (without WYSIWYG enabled) click on "Insert Variable" and insert any variable
3. Add a dollar sign "$" before this variable declaration, for instance ${{config path="web/unsecure/base_url"}}
4. Save the page

### Expected result

1. The page is saved and the user lands on the page listing and can see all pages listed

### Actual result

1. The page is saved and the user lands on the page listing but the spinner stays and pages do not appear

The same behaviour can be expected also with custom variables, for instance ${{customVar code=free_shipping_amount}}.
It is not specific to CMS pages, it can be experiences with CMS blocks too.

![capture d ecran 2017-08-11 a 10 01 58](https://user-images.githubusercontent.com/9135450/29216338-55169ee4-7e7c-11e7-86d6-a076f61ba92c.png)

**Comment by Skullsneeze** [comment]
I'm currently investigating the issue

**Comment by Skullsneeze** [comment]
Bug confirmed (against 2.2-develop) . Currently looking into a fix to resolve the issue

**Comment by slackerzz** [comment]
I'm working on it #SQUASHTOBERFEST

This issue affects also static blocks.
The problem here is that the string `${` is used in [template.js#L15](https://github.com/magento/magento2/blob/2.2-develop/lib/web/mage/utils/template.js#L15) as template syntax starting tag.
When `isTemplate` is called with the cms page/block content as arguments it returns a non zero value and the `render` function is called against it at [template.js#L174](https://github.com/magento/magento2/blob/2.2-develop/lib/web/mage/utils/template.js#L174) and we get:
```
Uncaught SyntaxError: Unexpected identifier
    at template (template.js:43)
    at render (template.js:97)
    at iterate (template.js:174)
    at Function._.each._.forEach (underscore.js:150)
    at iterate (template.js:176)
    at Function._.each._.forEach (underscore.js:150)
    at Object.template (template.js:161)
    at UiClass.initConfig (class.js:89)
    at UiClass.initialize (class.js:70)
    at UiClass._super (wrapper.js:106)
```

A possible solutions can be to escape in some way the cms page/block content before saving it. The point is how.
Simply replacing `${` with `$ {`? Alternatives?

**Comment by orlangur** [comment]
> Simply replacing `${` with `$ {`? Alternatives?

Why would someone ever use it in normal circumstances? it can be simply stripped or validated so that you cannot save such character combination.

**Comment by hostep** [comment]
> Why would someone ever use it in normal circumstances?

Perhaps someone wants to show a dollar currency symbol in front of a value which comes from a variable?

Other proposed solution: put add a 'zero width space' character (U+200B) in between them, or maybe even better a 'word joiner' character (U+2060), see: https://en.wikipedia.org/wiki/Whitespace_character#Unicode

Just tested the two above solutions, using a 'zero width space' character still fails, because it gets stripped out for some reason, but with the 'word joiner' character it seems to work, here's an example, you should be able to copy/paste it and it should work as expected:
```
$⁠{{config path="web/unsecure/base_url"}}
```

**Comment by Skullsneeze** [comment]
I've noticed that this not only goes for cms pages / block, but actually affects all text fields (product names, category titles, etc.) A solution like @hostep supplied might be a fix. I was currently working on a "global" validator that would notify the user that the "${" character combination is not allowed, but I'm thinking the fix mentioned before might be a better solution, since it still allows the desired output for the user

**Comment by slackerzz** [comment]
I was thinking about substituting `$` with `&#36;` which is the corresponding HTML Entity but the solution proposed by @hostep  seems better.

**Comment by orlangur** [comment]
Does template syntax like `${something}` WITHOUT backticks makes some sense actually?

According to https://github.com/magento/magento2/blob/2.2-develop/lib/web/mage/utils/template.js#L23 and https://lukehaas.me/articles/es6-in-a-nutshell/ it looks like not.

Thus proper fix should be applied to https://github.com/magento/magento2/blob/2.2-develop/lib/web/mage/utils/template.js#L79 so that it checks not only for `${` but also for backticks.

**Comment by slackerzz** [comment]
> Does template syntax like ${something} WITHOUT backticks makes some sense actually?

Yes, `${something}` is largely used in magento js without backticks.

This are not real ES6 template literals:
- https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Ui/view/base/web/js/grid/data-storage.js#L23
- https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/AdminNotification/view/adminhtml/web/js/grid/listing.js#L19
- ...and many more

`hasStringTmpls` at https://github.com/magento/magento2/blob/2.2-develop/lib/web/mage/utils/template.js#L22 tests if the browser supports ES6 template literal.
If true `template()` uses it at https://github.com/magento/magento2/blob/2.2-develop/lib/web/mage/utils/template.js#L43 

    return eval('`' + tmpl + '`');

If the browser doesn't support ES6 string templates `template()` uses the underscore templates renderer.

**Comment by orlangur** [comment]
True, too bad ES6 templates are not integrated natively and it cannot be even fixed easily due to BC constraints.

I like approach suggested by @hostep more as well, let's go this way, will do a full code review now.

**Comment by orlangur** [comment]
@slackerzz another option just came to my mind - can we maybe treat as template everything containing `${`, but not containing `${{`? Is `${{` a valid syntax for ES6 templates and whether it is used widely already in core?

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-81329

**Comment by sdzhepa** [comment]
Hello @fxdegroot and @slackerzz 

I found the issue was fixed 

Should be available for 2.3.0, see commits for more details:
- bac95762553a252c7f110a6b581b8b00c1ffcafd

This ticket can be closed.
Please, feel free to reopen or create a new one if issue still exists or was not fully fixed

Thank you for feedback and collaboration
